### ISSUE TITLE
Error pulling from gcloud private registry

### ISSUE THREAD

**Issue by alessioguglielmo** [spotify/docker-client#447]
## Description

statusCode returns code 200. However, next pull give me an error. In gcloud private registry I have an image tagged as red:latest
## How to reproduce

```
final DockerClient docker = DefaultDockerClient.builder()
                .uri(URI.create("http://192.168.1.8:2375"))             
                .build();
AuthConfig authConfig = null;

            try {
                authConfig = AuthConfig.fromDockerConfig("https://gcr.io").build();
            } catch (IOException e1) {
                // TODO Auto-generated catch block
                e1.printStackTrace();
            }


        final int statusCode = docker.auth(authConfig);
        System.out.println("Auth esito: " + statusCode);

        docker.pull("gcr.io/pure-muse-748/red:latest", authConfig);

        System.out.println("pull eseguito.");
```
## What do you expect

Pull of private image on my local Docker host
## What happened instead

```
ERRO[0185] Attempting next endpoint for pull after error: manifest unknown: Failed to fetch "latest" from request "/v2/pure-muse-748/red/manifests/latest". 

ERRO[0186] Not continuing with pull after error: Error: Status 403 trying to pull repository pure-muse-748/red: "Unable to access the repository: pure-muse-748/red; please verify that it exists and you have permission to access it (no valid credential was supplied)." 

```
## Software:

```
docker -H :2375 version
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Wed Apr 27 00:34:42 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   5604cbe
 Built:        Wed Apr 27 00:34:42 2016
 OS/Arch:      linux/amd64
```
## Full backtrace

```
Exception in thread "main" java.lang.reflect.InvocationTargetException
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.eclipse.jdt.internal.jarinjarloader.JarRsrcLoader.main(JarRsrcLoader.java:58)
Caused by: com.spotify.docker.client.exceptions.ImagePullFailedException: Image pull failed: gcr.io/pure-muse-748/red:latest: ProgressMessage{id=null, status=null, stream=null, error=Error: Status 403 trying to pull repository pure-muse-748/red: "Unable to access the repository: pure-muse-748/red; please verify that it exists and you have permission to access it (no valid credential was supplied).", progress=null, progressDetail=null}
    at com.spotify.docker.client.LoggingPullHandler.progress(LoggingPullHandler.java:44)
    at com.spotify.docker.client.ProgressStream.tail(ProgressStream.java:77)
    at com.spotify.docker.client.DefaultDockerClient.pull(DefaultDockerClient.java:987)
    at com.spotify.docker.client.DefaultDockerClient.pull(DefaultDockerClient.java:967)
    at client.Main.main(Main.java:54)
    ... 5 more
```

**Comment by mattnworb** [comment]
What does your `~/.docker/config.json` look like?

**Comment by mattnworb** [comment]
To be clear, it's not necessary to post the entire contents of the file - but it would be good to verify the structure of the file and that it has an entry for `"https://gcr.io"`

**Comment by alessioguglielmo** [comment]
Strange thing is that first authentication returns code 200
This is my ~/.docker/config.json 

`{
	"auths": {
		"https://gcr.io": {
			"auth": "X2pzb25fa2V5OnsKICAidHlwZSI6ICJzZXJ2aWNlX2FjY291bnQiLAogICJwcm9qZWN0X2lkIjogInB1c
[...etc] [... etc]
mVsb3Blci5nc2VydmljZWFjY291bnQuY29tIgp9"
		}
	}
}`

**Comment by johnflavin** [comment]
I could be wrong, but I think authenticating with a token isn't supported by `docker-client` yet. That was added in docker 1.10, and we haven't caught up with all the 1.10 API features yet.

In the meantime you could try authenticating with a username and password instead. But I don't know the structure of the `~/.docker/config.json` file, so I don't know exactly how you would need to put them in there.

**Comment by mattnworb** [comment]
@johnflavin you are right, the AuthConfig.Builder class expects that the `auth` value can be base64 decoded and then splits it on `:` to find a username and password to use.

**Comment by alessioguglielmo** [comment]
I tried to make a `docker login` and authenticate with user + pass but config.json now has 
```
{
	"auths": {
		"https://gcr.io": {
			"auth": "b2F1dGgyYWNjZ ... ... X1BIUA=="
		}
	}
}
```
same 403 error appears
I haven't found a valid config template example around in the web :(

**Comment by mattnworb** [comment]
Coming back to this issue, I think there are a few things going on:

1. docker-client doesn't know how to use tokens as @johnflavin 

2. The access tokens for Google Container Registry (what the docker config file gets populated with if you run `gcloud docker -a) are short-lived tokens that are only valid for some period of time

    For instance if I run `gcloud docker --authorize-only` to configure the daemon, the output is pretty explicit:

    ```
    $> gcloud docker -a
    Short-lived access for ['gcr.io', 'us.gcr.io', 'eu.gcr.io', 'asia.gcr.io', 'b.gcr.io', 'bucket.gcr.io', 'l.gcr.io', 'launcher.gcr.io', 'appengine.gcr.io', 'us-mirror.gcr.io', 'eu-mirror.gcr.io', 'asia-mirror.gcr.io', 'mirror.gcr.io'] configured.
    ```

**Comment by mattnworb** [comment]
Another option is to use https://github.com/GoogleCloudPlatform/docker-credential-gcr, which allows for integrating with GCR/gcloud via [the somewhat-newish support for Credentials Helpers in docker](https://docs.docker.com/engine/reference/commandline/login/#credentials-store).

**Comment by mattnworb** [comment]
Actually I don't think that a docker credential helper would work, since the authentication tokens are expected to be obtained by the client (e.g. the `docker` CLI) and the daemon / Remote API doesn't know anything about those.

**Comment by mattnworb** [comment]
addressed in docker-client v8.6.0 and later versions
